Show an error message if the browser does not support WritableStream

fridfinnurm opened the pull request:
I think the title says it all :)


I'm `AverageFedora` on reddit, if you want to get in touch.

fridfinnurm committed:
Added an error message in case the browser doesn't support streaming^C

TejasQ commented on examples/streaming-pwa/css/main.src.css:
```diff
@@ -139,3 +139,10 @@ button:hover {
   text-align: left;
 }
+
+.error-message {
```
Nice!

TejasQ commented on examples/streaming-pwa/js/index.js:
```diff
@@ -5,6 +5,24 @@ const searchField = document.querySelector("#search-field")
+if( window.WritableStream === undefined ){
+  
+  // oh no
+  console.log("Oops! Your current browser doesn't support streaming :(")
```
`console.error` may be better here?

fridfinnurm replied:
Absolutely, just slipped my mind

TejasQ commented on examples/streaming-pwa/js/index.js:
```diff
@@ -5,6 +5,24 @@ const searchField = document.querySelector("#search-field")
 const resultsContainer = document.querySelector("#results")
 
+// Quit if the browser doesn't support WritableStream
+if( window.WritableStream === undefined ){
```
It maybe better to use the `typeof` check here:
`if (typeof window.ReadableStream === 'undefined' && typeof window.WritableStream === 'undefined')`

TejasQ replied:
Also accounting for Readable AND writable.

fridfinnurm replied:
I believe we can skip `typeof` here because we're getting `WritableStream` as property of the window object. the `typeof` check is useful because `typeof x` does not throw an error if x is undefined, while `x === undefined` throws a reference error. In this case, by referring to `WritableStream` as a property of `window` we can avoid this error, as trying to get an undefined property of an object just returns undefined e.g. `foo = {}; foo.bar === undefined // True`

We could absolutely write this as `typeof ReadableStream === "undefined" || typeof WritableStream === "undefined"`, and it's really just a matter of style. After looking at it, I personally prefer the typeof version.

The reason I only checked `WritableStream` is because no browser spec currently implements Writable without also implementing Readable (Firefox actually implements only Readable), but more verbosity won't hurt :)

In fact, checking `if (typeof window.ReadableStream === 'undefined' && typeof window.WritableStream === 'undefined')` would actually malfunction in firefox, due to the weirdness I mentioned.

TejasQ replied:
Man, thanks for the comment. I was definitely able to learn from it. Sounds good.

TejasQ commented on examples/streaming-pwa/js/index.js:
```diff
@@ -5,6 +5,24 @@ const searchField = document.querySelector("#search-field")
+  console.log("Oops! Your current browser doesn't support streaming :(")
+
+  // Alert the user via dangerously set inner html
+  resultsContainer.innerHTML = `
```
I would actually put this in the `index.html` and then just add a CSS class `show` with JS to the querySelected element, for encapsulation. Do you think that makes sense?

fridfinnurm replied:
Just leave a `<div class="hidden error">...</div>` in the index file, then get the component by the .error identifier and set className="error" (no need for a special show class, just a hide {display: none; })? That sounds reasonable

TejasQ replied:
Or even `document.querySelector('.error').classList.remove('hidden')` to show it.

TejasQ replied:
Ideally the querySelector call lives at the top with the other querySelectors and we simply do `error.classList.remove('hidden') // <- show the error`, following the `Basically,` style of simple communication.

fridfinnurm replied:
Sounds good

TejasQ replied:
Thanks for the discussion! Excited for the next commit. 😉

TejasQ commented on examples/streaming-pwa/js/index.js:
```diff
@@ -5,6 +5,24 @@ const searchField = document.querySelector("#search-field")
+  `
+
+  // Stop execution
+  throw new Error("Browser does not support streaming");
```
We can skip the `console.log` entirely on line 11 and just do this! Genius!

fridfinnurm replied:
Yeah, printing and having an error logged does seem a bit redundant now that I think about it heh

TejasQ reviewed (changes requested):
Thanks _so much_ for your contribution! Care to have a discussion about it? 😄

fridfinnurm committed:
Addressed points mentioned in PR review.

Check both Readable and Writable -Stream when cheking if browser supports streaming.
Embed error message in App.jsx as a hidden div.
Recompiled css.

TejasQ reviewed (approved):
Awesome! Thanks so much!
